sympy/simplify/_cse_diff.py

```diff
@@ -0,0 +1,244 @@
+    which are useful in the substitution process.
+
+    NOTE: The output Jacobian might not be always exactly in a DAG form, but close to it.
+
+    Parameters
+    ==========
+
+    replacements : list
+        A list of tuples containing the CSE replacement symbols and their corresponding
+        expressions.
+
+    reduced_expr : list
+        The reduced expression after the CSE operation.
+
+    wrt : Matrix
+        The matrix of expressions with respect to which to differentiate the reduced
+        expression.
+
+    """
+
+    if not isinstance(reduced_expr[0], MatrixBase):
+        raise TypeError("``expr`` must be of matrix type")
+
+    if not (reduced_expr[0].shape[0] == 1 or reduced_expr[0].shape[1] == 1):
+        raise TypeError("``expr`` must be a row or a column matrix")
+
+    if not isinstance(wrt, (MatrixBase, list, tuple)):
+        raise TypeError("``wrt`` must be an iterable of variables")
+
+    elif isinstance(wrt, (list, tuple)):
+        wrt = Matrix(wrt)
+
+    if not (wrt.shape[0] == 1 or wrt.shape[1] == 1):
+        raise TypeError("``wrt`` must be a row or a column matrix")
+
+    replacements, reduced_expr = _postprocess(replacements, reduced_expr)
+
+    if replacements:
+        rep_sym, sub_expr = map(Matrix, zip(*replacements))
+    else:
+        rep_sym, sub_expr = Matrix([]), Matrix([])
+
+    l_sub, l_wrt, l_red = len(sub_expr), len(wrt), len(reduced_expr[0])
+
+    f1 = {
+        (i, j): diff_value
+        for i, r in enumerate(reduced_expr[0])
+        for j, w in enumerate(wrt)
+        if (diff_value := r.diff(w)) != 0
+    }
+
+    if not replacements:
+        return [], SparseMatrix(l_red, l_wrt, f1), []
+
+    f2 = {
+        (i, j): diff_value
+        for i, (r, fs) in enumerate([(r, r.free_symbols) for r in reduced_expr[0]])
+        for j, s in enumerate(rep_sym)
+        if s in fs and (diff_value := r.diff(s)) != 0
+    }
+
+    precomputed_fs = [
+        {symbol for symbol in s.free_symbols if re.compile(r'x\d+').fullmatch(symbol.name)}
+        for s in sub_expr
+    ]
+
+    C = Counter({(0, j): diff_value for j, w in enumerate(wrt) if (diff_value := sub_expr[0].diff(w)) != 0})
+
+    for i in range(1, l_sub):
+        Bi = {(i, j): diff_value for j in range(i + 1)
+              if rep_sym[j] in precomputed_fs[i] and (diff_value := sub_expr[i].diff(rep_sym[j])) != 0}
+
+        Ai = Counter({(i, j): diff_value for j, w in enumerate(wrt)
+                      if (diff_value := sub_expr[i].diff(w)) != 0})
+
+        if Bi:
+            Ci = _dok_matrix_multiply(Bi, C)
+            Ci.update(Ai)  # Use Counter's update method to add Ai to Ci
+            C.update(Ci)  # Update C with the result
+        else:
+            C.update(Ai)
+
+    J = _dok_matrix_multiply(f2, C)
+
+    for (i, j), value in f1.items():
+        J[(i, j)] += value
+
+    J = SparseMatrix(l_red, l_wrt, J)
+    J = reduced_expr[0].__class__(J)
+
+    return replacements, J, precomputed_fs
```

Comment:
Add unit tests to make sure these outputs are structured same as cse() (those outputs are a bit complex I think and it should try to follow same pattern)